cmd/juju/controller/register.go

```diff
@@ -623,3 +632,40 @@ type byteAtATimeReader struct {
 func (r byteAtATimeReader) Read(out []byte) (int, error) {
 	return r.Reader.Read(out[:1])
 }
+
+// ensureNotKnownEndpoint checks whether any controllers in the local client
+// cache contain the provided endpoint and returns an error if that is the
+// case.
+func ensureNotKnownEndpoint(store jujuclient.ClientStore, endpoint string) error {
+	existingDetails, existingName, err := store.ControllerByAPIEndpoints(endpoint)
+	if err != nil && !errors.IsNotFound(err) {
+		return errors.Trace(err)
+	}
+
+	if existingDetails == nil {
+		return nil
+	}
+
+	// Check if we know the username for this controller
+	accountDetails, err := store.AccountDetails(existingName)
+	if err != nil && !errors.IsNotFound(err) {
+		return errors.Trace(err)
+	}
+
+	if accountDetails != nil {
+		var buf bytes.Buffer
+		if err := alreadyRegisteredMessageT.Execute(
+			&buf,
+			map[string]interface{}{
```

Comment:
I'd rather we used a struct here, so we keep type some type safety by removing the stingified keys in the map.

Reply:
Pushed https://github.com/juju/juju/pull/10204/commits/db81428327442f0f21af8286f45bac1f6abe9dfe to address this comment.